File: frontend/src/services/transactionService.js
import apiClient from './apiClient';
import { TRANSACTION_ENDPOINTS } from '../constants/apiEndpoints';

// src/services/transactionService.js
const transactionService = {
  transferFunds: async (transferData) => {
    // RETURNS: { statusCode, message, data: { transactionId, ... } }
    const response = await apiClient.post(TRANSACTION_ENDPOINTS.TRANSFER_FUNDS, transferData);
    return response.data; 
  },
  
  getAccountTransactions: async (accountId, page = 0, size = 10) => {
    // RETURNS: { statusCode, message, data: { content: [...], totalElements: 10 } }
    const response = await apiClient.get(TRANSACTION_ENDPOINTS.GET_ACCOUNT_TRANSACTIONS(accountId), {
      params: { page, size }
    });
    return response.data;
  },
  // ... other methods


  // Get transaction by ID
  getTransactionById: async (transactionId) => {
    try {
      const response = await apiClient.get(TRANSACTION_ENDPOINTS.GET_TRANSACTION(transactionId));
      return response.data;
    } catch (error) {
      throw error.response?.data || error;
    }
  },

  // Get transaction by reference
  getTransactionByReference: async (reference) => {
    try {
      const response = await apiClient.get(
        TRANSACTION_ENDPOINTS.GET_TRANSACTION_BY_REFERENCE(reference)
      );
      return response.data;
    } catch (error) {
      throw error.response?.data || error;
    }
  },

  // Get all transactions (Admin)
  getAllTransactions: async (page = 0, size = 10) => {
    try {
      const response = await apiClient.get(TRANSACTION_ENDPOINTS.GET_ALL_TRANSACTIONS, {
        params: { page, size },
      });
      return response.data;
    } catch (error) {
      throw error.response?.data || error;
    }
  },

  // Get account transactions
  getAccountTransactions: async (accountId, page = 0, size = 10) => {
    try {
      const response = await apiClient.get(
        TRANSACTION_ENDPOINTS.GET_ACCOUNT_TRANSACTIONS(accountId),
        { params: { page, size } }
      );
      return response.data;
    } catch (error) {
      throw error.response?.data || error;
    }
  },

  // Get outgoing transactions
  getOutgoingTransactions: async (accountId, page = 0, size = 10) => {
    try {
      const response = await apiClient.get(
        TRANSACTION_ENDPOINTS.GET_OUTGOING_TRANSACTIONS(accountId),
        { params: { page, size } }
      );
      return response.data;
    } catch (error) {
      throw error.response?.data || error;
    }
  },

  // Get incoming transactions
  getIncomingTransactions: async (accountId, page = 0, size = 10) => {
    try {
      const response = await apiClient.get(
        TRANSACTION_ENDPOINTS.GET_INCOMING_TRANSACTIONS(accountId),
        { params: { page, size } }
      );
      return response.data;
    } catch (error) {
      throw error.response?.data || error;
    }
  },

  // Get transaction history
  getTransactionHistory: async (accountId, page = 0, size = 10) => {
    try {
      const response = await apiClient.get(
        TRANSACTION_ENDPOINTS.GET_TRANSACTION_HISTORY(accountId),
        { params: { page, size } }
      );
      return response.data;
    } catch (error) {
      throw error.response?.data || error;
    }
  },

  // Reverse transaction
  reverseTransaction: async (transactionId) => {
    try {
      const response = await apiClient.post(
        TRANSACTION_ENDPOINTS.REVERSE_TRANSACTION(transactionId)
      );
      return response.data;
    } catch (error) {
      throw error.response?.data || error;
    }
  },

  // Get total outgoing amount
  getTotalOutgoing: async (accountId) => {
    try {
      const response = await apiClient.get(
        TRANSACTION_ENDPOINTS.GET_TOTAL_OUTGOING(accountId)
      );
      return response.data;
    } catch (error) {
      throw error.response?.data || error;
    }
  },
};

export default transactionService;

File: backend/src/main/java/com/snlbanking/service/TransactionService.java
package com.snlbanking.service;

import com.snlbanking.dto.FundTransferDTO;
import com.snlbanking.entity.Transaction;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * Transaction Service Interface
 * Defines business logic operations for Transaction entity
 * 
 * @author SNL Banking Team
 * @version 1.0.0
 */
public interface TransactionService {

    /**
     * Transfer funds between accounts
     */
    Transaction transferFunds(FundTransferDTO transferDTO);

    /**
     * Get transaction by ID
     */
    Transaction getTransactionById(Long transactionId);

    /**
     * Get transaction by reference
     */
    Transaction getTransactionByReference(String transactionReference);

    /**
     * Get all transactions for an account
     */
    Page<Transaction> getAccountTransactions(Long accountId, Pageable pageable);

    /**
     * Get outgoing transactions for an account
     */
    Page<Transaction> getOutgoingTransactions(Long accountId, Pageable pageable);

    /**
     * Get incoming transactions for an account
     */
    Page<Transaction> getIncomingTransactions(Long accountId, Pageable pageable);

    /**
     * Get transactions within date range
     */
    Page<Transaction> getTransactionsByDateRange(Long accountId, LocalDateTime startDate, 
                                                 LocalDateTime endDate, Pageable pageable);

    /**
     * Get transactions by type
     */
    Page<Transaction> getTransactionsByType(Transaction.TransactionType type, Pageable pageable);

    /**
     * Get transactions by status
     */
    Page<Transaction> getTransactionsByStatus(Transaction.TransactionStatus status, Pageable pageable);

    /**
     * Calculate total outgoing amount for an account
     */
    BigDecimal calculateTotalOutgoing(Long accountId);

    /**
     * Reverse a transaction
     */
    void reverseTransaction(Long transactionId);

    /**
     * Get all transactions with pagination
     */
    Page<Transaction> getAllTransactions(Pageable pageable);
}
